src/hotspot/share/gc/g1/g1Policy.cpp

```diff
@@ -922,26 +928,44 @@ void G1Policy::record_young_collection_end(bool concurrent_operation_is_full_mar
   }
 
   // Note that _mmu_tracker->max_gc_time() returns the time in seconds.
-  double scan_logged_cards_time_goal_ms = _mmu_tracker->max_gc_time() * MILLIUNITS * G1RSetUpdatingPauseTimePercent / 100.0;
+  double logged_cards_time_goal_ms = _mmu_tracker->max_gc_time() * MILLIUNITS * G1RSetUpdatingPauseTimePercent / 100.0;
 
-  if (scan_logged_cards_time_goal_ms < merge_hcc_time_ms) {
+  if (logged_cards_time_goal_ms < merge_hcc_time_ms) {
     log_debug(gc, ergo, refine)("Adjust concurrent refinement thresholds (scanning the HCC expected to take longer than Update RS time goal)."
                                 "Logged Cards Scan time goal: %1.2fms Scan HCC time: %1.2fms",
-                                scan_logged_cards_time_goal_ms, merge_hcc_time_ms);
+                                logged_cards_time_goal_ms, merge_hcc_time_ms);
 
-    scan_logged_cards_time_goal_ms = 0;
+    logged_cards_time_goal_ms = 0;
   } else {
-    scan_logged_cards_time_goal_ms -= merge_hcc_time_ms;
+    logged_cards_time_goal_ms -= merge_hcc_time_ms;
   }
 
-  double const logged_cards_time = logged_cards_processing_time();
+  double const logged_cards_time_ms = logged_cards_processing_time();
+  size_t logged_cards =
+    phase_times()->sum_thread_work_items(G1GCPhaseTimes::MergeLB,
+                                         G1GCPhaseTimes::MergeLBDirtyCards);
+  size_t hcc_cards =
+    phase_times()->sum_thread_work_items(G1GCPhaseTimes::MergeHCC,
+                                         G1GCPhaseTimes::MergeHCCDirtyCards);
+  bool exceeded_goal = logged_cards_time_goal_ms < logged_cards_time_ms;
+  size_t predicted_thread_buffer_cards = _analytics->predict_dirtied_cards_in_thread_buffers();
+  G1ConcurrentRefine* cr = _g1h->concurrent_refine();
 
-  log_debug(gc, ergo, refine)("Concurrent refinement times: Logged Cards Scan time goal: %1.2fms Logged Cards Scan time: %1.2fms HCC time: %1.2fms",
-                              scan_logged_cards_time_goal_ms, logged_cards_time, merge_hcc_time_ms);
+  log_debug(gc, ergo, refine)
+           ("GC refinement: goal: %zu + %zu / %1.2fms, actual: %zu / %1.2fms, HCC: %zu / %1.2fms%s",
+            cr->pending_cards_target(),
+            predicted_thread_buffer_cards,
+            logged_cards_time_goal_ms,
+            logged_cards,
+            logged_cards_time_ms,
+            hcc_cards,
+            merge_hcc_time_ms,
+            (exceeded_goal ? " (exceeded goal)" : ""));
```

Comment:
Oh, I forgot about that.  I was going to try to come up with something better there, like somehow printing "unlimited" or something like that.

Reply:
Stefan and I talked about this offline, and decided to leave it as is for now.
- Printing a string like "unlimited" for that case makes parsing more complicated.
- Printing 0 can be confusing because that's a valid value.  Neither of us likes log messages that contain "lies".
- Suppressing the message in this case loses the information used to calculate the first normal value.